sharing_portal/views.py

```diff
@@ -1,226 +1,180 @@
 
 from .forms import (
     ArtifactForm,
-    ArtifactVersionForm,
     AuthorFormset,
     ShareArtifactForm,
-    ZenodoPublishFormset,
     RequestDaypassForm,
     ReviewDaypassForm,
 )
-from .models import Artifact, ArtifactVersion, Author, ShareTarget, DaypassRequest
-from .tasks import publish_to_zenodo
+from .models import DaypassRequest, DaypassProject
+from . import trovi
 from projects.views import (
     add_project_invitation,
     get_invite_url,
     get_project_membership_managers,
     is_membership_manager,
     manage_membership_in_scope,
 )
+from .zenodo import ZenodoClient
+
+from urllib.parse import urlencode
 
 import logging
+from datetime import datetime, timedelta
 
 LOG = logging.getLogger(__name__)
 
 SHARING_KEY_PARAM = "s"
 
 
-class ArtifactSerializer(serializers.ModelSerializer):
-    class Meta:
-        model = Artifact
-        fields = "__all__"
+def ensure_trovi_token(view_func):
+    def _wrapped_view(request, *args, **kwargs):
+        if (
+            request.session.get("trovi_token_expiration")
+            and datetime.now().timestamp() > request.session["trovi_token_expiration"]
+        ):
+            del request.session["trovi_token_expiration"]
+            del request.session["trovi_token"]
+
+        if not request.session.get("trovi_token"):
+            if request.session.get("oidc_access_token"):
+                try:
+                    response = trovi.get_token(
+                        request.session.get("oidc_access_token"),
+                        is_admin=request.user.is_staff,
+                    )
+                    request.session["trovi_token"] = response["access_token"]
+                    request.session["trovi_token_expiration"] = min(
+                        (
+                            datetime.now() + timedelta(seconds=response["expires_in"])
+                        ).timestamp(),
+                        request.session["oidc_token_expiration"],
+                    )
+                except trovi.TroviException:
+                    LOG.error("Error getting trovi token")
+            else:
+                # Set an empty token
+                request.session["trovi_token"] = ""
+        return view_func(request, *args, **kwargs)
+
+    return _wrapped_view
 
 
 def can_edit(request, artifact):
-    if artifact.created_by == request.user:
-        return True
-    if request.user.is_staff:
-        return True
-    return False
+    return _owns_artifact(request.user, artifact) or request.user.is_staff
 
 
 def check_edit_permission(func):
     def wrapper(request, *args, **kwargs):
-        pk = kwargs.pop('pk')
-        artifact = get_object_or_404(Artifact, pk=pk)
+        pk = kwargs.pop("pk")
+        try:
+            artifact = trovi.get_artifact_by_trovi_uuid(
+                request.session.get("trovi_token"), pk
+            )
+        except trovi.TroviException:
+            raise Http404("That artifact does not exist, or is private")
```

Comment:
Might be good to check trovi's response code to make sure it's 404. We would want to log 500s so that we know when trovi is being bad.